### PR TITLE
Add CTC binary and switch controls

### PR DESCRIPTION
**Pull request by eddy1021**
Add enablement and params controls of Google CTC component.

### CODE REVIEW

File: src/audio/google/google_ctc_audio_processing.c [hunk 1]
```diff
@@ -279,7 +300,7 @@ static int google_ctc_audio_processing_reconfigure(struct processing_module *mod
 	struct comp_dev *dev = mod->dev;
 	uint8_t *config;
 	size_t size;
-	int ret;
+	int ret = 0;
```
Comment (lyakh) [inline src/audio/google/google_ctc_audio_processing.c:303]: not needed

File: src/audio/google/google_ctc_audio_processing.c [hunk 2]
```diff
@@ -80,6 +79,12 @@ static void ctc_s16_default(struct google_ctc_audio_processing_comp_data *cd,
 	int samples_to_written = MIN(samples, audio_stream_samples_without_wrap_s16(sink, dest));
 	int written_samples = 0;
 
+	if (!cd->enabled) {
```
Comment (cujomalainey) [inline src/audio/google/google_ctc_audio_processing.c:82]: this should be split into a different commit and using the switch control

File: src/audio/google/google_ctc_audio_processing.c [hunk 3]
```diff
@@ -295,32 +316,21 @@ static int google_ctc_audio_processing_reconfigure(struct processing_module *mod
 	}
 
 	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
-		  config, size);
+		  cd->config, size);
 
-	cd->reconfigure = false;
+	cd->config = (struct google_ctc_config *)config;
+	if (cd->config)
```
Comment (lyakh) [inline src/audio/google/google_ctc_audio_processing.c:322]: this check is redundant - `config` has been checked to be non-`NULL` 9 lines above

File: src/audio/google/google_ctc_audio_processing.c [hunk 4]
```diff
@@ -295,32 +316,21 @@ static int google_ctc_audio_processing_reconfigure(struct processing_module *mod
 	}
 
 	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
-		  config, size);
+		  cd->config, size);
```
Comment (lyakh) [inline src/audio/google/google_ctc_audio_processing.c:319]: is this correct - before the assignment?

File: src/audio/google/google_ctc_audio_processing.c [hunk 5]
```diff
@@ -408,6 +431,7 @@ static int ctc_process(struct processing_module *mod,
 	}
 
 	cd->ctc_func(cd, source, sink, &input_buffers[0], &output_buffers[0], frames);
+
```
Comment (lyakh) [inline src/audio/google/google_ctc_audio_processing.c:434]: drive-by, and although in general I personally do prefer an empty line before `return`, this is inconsistent with your above changes, where there's no such line

File: src/audio/google/google_ctc_audio_processing.c (hunk 2 above)
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing.c:82]: Split it into another commit now. Thanks!

File: src/audio/google/google_ctc_audio_processing_ipc3.c [hunk 6]
```diff
@@ -39,6 +39,13 @@ int ctc_set_config(struct processing_module *mod, uint32_t param_id,
 			cd->reconfigure = true;
 		}
 		return 0;
+	case SOF_CTRL_CMD_SWITCH:
+		if (cdata->num_elems == 1) {
+			cd->enabled = cdata->chanv[0].value;
+			comp_info(dev, "ctc_set_config(), enabled = %d",
+				  cd->enabled);
+			return 0;
+		}
```
Comment (lyakh) [inline src/audio/google/google_ctc_audio_processing_ipc3.c:48]: is this intended - to fall through here? If so, please use `COMPILER_FALLTHROUGH`

**Review by lyakh — CHANGES_REQUESTED** [review]
I think there's a high likelihood that the fall-through wasn't intended

File: src/audio/google/google_ctc_audio_processing_ipc3.c (hunk 6 above)
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing_ipc3.c:53]: It was intended to fall through to `default` and returned as invalid control.
I changed to handle in a different error message to make it clear and more error-proof.

Thanks for pointing this out!

Comment (lyakh) [reply src/audio/google/google_ctc_audio_processing_ipc3.c:53]: right, this fixes it, thanks. Just a soft wish - in such cases SOF usually prefers no `else`. I.e.
```
if (x) {
    handle_x();
    return 0;
}
print("error");
return -Esuitableerror;
```
or perhaps
```
if (e) {
    print("error");
    return -Esuitableerror;
}
// proceed normally
```

File: src/audio/google/google_ctc_audio_processing.c [hunk 7]
```diff
@@ -243,6 +243,8 @@ static int ctc_free(struct processing_module *mod)
 {
 	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
 
+	comp_info(mod->dev, "ctc_free()");
```
Comment (kv2019i) [inline src/audio/google/google_ctc_audio_processing.c:246]: This 3rd commit could have been a separate PR as well. Not a blocker.

**Review by kv2019i — APPROVED** [review]

File: src/audio/google/google_ctc_audio_processing_ipc4.c [hunk 8]
```diff
@@ -25,16 +25,26 @@ int ctc_set_config(struct processing_module *mod, uint32_t param_id,
...
 	case SOF_IPC4_SWITCH_CONTROL_PARAM_ID:
+		if (ctl->id == 0 && ctl->num_elems == 1) {
+			cd->enabled = ctl->chanv[0].value;
+			comp_info(mod->dev, "ctc_set_config(): enabled = %d", cd->enabled);
+		} else {
+			comp_err(mod->dev, "ctc_set_config(): Illegal control id = %d, num_elems = %d",
+				 ctl->id, ctl->num_elems);
+			return -EINVAL;
+		}
+		return 0;
```
Comment (cujomalainey) [inline src/audio/google/google_ctc_audio_processing_ipc4.c:44]: ditto to @kv2019i comments. can this be cleaned to single branch?
```
		if (ctl->id == 0 && ctl->num_elems == 1) {
			cd->enabled = ctl->chanv[0].value;
			comp_info(mod->dev, "ctc_set_config(): enabled = %d", cd->enabled);
                        return 0;
		} 
		comp_err(mod->dev, "ctc_set_config(): Illegal control id = %d, num_elems = %d",
				 ctl->id, ctl->num_elems);
		return -EINVAL;
```

File: src/audio/google/google_ctc_audio_processing.h [hunk 9]
```diff
@@ -37,10 +37,18 @@ struct google_ctc_audio_processing_comp_data {
...
+	struct google_ctc_config *config;
 	bool reconfigure;
 	ctc_func ctc_func;
 };
 
+struct google_ctc_config {
+	uint32_t size;
+
+	/* reserved */
+	uint32_t reserved[4];
```
Comment (cujomalainey) [inline src/audio/google/google_ctc_audio_processing.h:50]: data[] at the end

File: src/audio/google/google_ctc_audio_processing_ipc3.c [hunk 10]
```diff
@@ -35,7 +35,7 @@ int ctc_set_config(struct processing_module *mod, uint32_t param_id,
 		if (ret)
 			return ret;
 		if (comp_is_new_data_blob_available(cd->tuning_handler)) {
-			comp_get_data_blob(cd->tuning_handler, NULL, NULL);
+			cd->config = comp_get_data_blob(cd->tuning_handler, NULL, NULL);
```
Comment (cujomalainey) [inline src/audio/google/google_ctc_audio_processing_ipc3.c:38]: now that you have the size, you should be doing a size check from the blob handler against the size in the header. Same for ipc4

File: src/audio/google/google_ctc_audio_processing_ipc4.c (hunk 8 above)
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing_ipc4.c]: done.

File: src/audio/google/google_ctc_audio_processing.h (hunk 9 above)
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing.h:50]: Added.

File: src/audio/google/google_ctc_audio_processing_ipc3.c (hunk 10 above)
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing_ipc3.c]: Thanks for catching this!

File: src/audio/google/google_ctc_audio_processing.h [hunk 11]
```diff
@@ -37,10 +37,23 @@ struct google_ctc_audio_processing_comp_data {
 	uint32_t chunk_frames;
 	GoogleCtcAudioProcessingState *state;
 	struct comp_data_blob_handler *tuning_handler;
+	struct google_ctc_config *config;
 	bool reconfigure;
 	ctc_func ctc_func;
 };
 
+struct google_ctc_config {
+	uint32_t size;
```
Comment (cujomalainey) [inline src/audio/google/google_ctc_audio_processing.h:47]: please define if this is inclusive or just the size of `data`

Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing.h:48]: done. Thanks!

File: src/audio/google/google_ctc_audio_processing_ipc3.c [hunk 12]
```diff
@@ -28,14 +28,21 @@ int ctc_set_config(struct processing_module *mod, uint32_t param_id,
...
 
 	switch (cdata->cmd) {
 	case SOF_CTRL_CMD_BINARY:
 		ret = comp_data_blob_set_cmd(cd->tuning_handler, cdata);
 		if (ret)
 			return ret;
 		if (comp_is_new_data_blob_available(cd->tuning_handler)) {
-			comp_get_data_blob(cd->tuning_handler, NULL, NULL);
+			cd->config = comp_get_data_blob(cd->tuning_handler, &size, NULL);
+			if (size != CTC_BLOB_CONFIG_SIZE) {
```
[the remark this answers is not in the document]
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing_ipc3.c:40]: Done.

**Review by cujomalainey — APPROVED** [review]

File: src/audio/google/google_ctc_audio_processing.c [hunk 13]
```diff
@@ -372,6 +357,7 @@ static int ctc_prepare(struct processing_module *mod,
 							     /*config=*/NULL,
 							     /*config_size=*/0);
 
+	cd->config = comp_get_data_blob(cd->tuning_handler, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (johnylin76) [reply src/audio/google/google_ctc_audio_processing.c:396]: It looks that on `prepare()` CTC is started without any config. So does that mean CTC will always apply config via `reconfigure()`?
As I can figure out, `cd->reconfigure` is toggled up by `ctc_set_config()` only, i.e. a "new" config blob is set explicitly. IIUC, it will not be toggled up in the following cases:
1. CTC gets restarted (`reset()` and then `prepare()`) --> user pauses playback > 10 seconds 
2. config blob is set to CTC while CTS is idle --> depends on start-up sequence\
(reference: https://github.com/thesofproject/sof/blob/main/src/audio/data_blob.c#L538)

I think 2 is easy to reproduce by setting CTC enabled prior to starting playback.

File: src/audio/google/google_ctc_audio_processing_ipc3.c [hunk 14]
```diff
@@ -52,9 +52,20 @@ int ctc_set_config(struct processing_module *mod, uint32_t param_id,
 			cd->reconfigure = true;
 		}
 		return 0;
+	case SOF_CTRL_CMD_SWITCH:
```
Comment (johnylin76) [inline src/audio/google/google_ctc_audio_processing_ipc3.c:55]: how about getter for switch control?

**Review by johnylin76 — COMMENTED** [review]
Please check concerns about the data blob setting process as I stated in comments. Thanks

File: src/audio/google/google_ctc_audio_processing.c [hunk 15]
```diff
@@ -297,32 +296,18 @@ static int google_ctc_audio_processing_reconfigure(struct processing_module *mod
 	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
 		  config, size);
 
-	cd->reconfigure = false;
+	cd->config = (struct google_ctc_config *)config;
```
[the remark this answers is not in the document]
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing.c]: Thanks for catching this!

I changed to the first implementation. It's much cleaner now.

File: src/audio/google/google_ctc_audio_processing.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing.c]: Changed to get the config before creating a CTC state then we can pass in a CTC config.

File: src/audio/google/google_ctc_audio_processing_ipc3.c (hunk 14 above)
Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing_ipc3.c:56]: Not really sure what getter do you mean.
Switch control is not going through the config blob.

Comment (cujomalainey) [reply src/audio/google/google_ctc_audio_processing_ipc3.c:56]: I don't think we need one, its not data that needs to be read out, if there is a case in the future then sure.

Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing_ipc3.c:56]: Thanks! added the getter now.

File: src/audio/google/google_ctc_audio_processing.c [hunk 16]
```diff
@@ -367,10 +352,16 @@ static int ctc_prepare(struct processing_module *mod,
 	}
 	cd->next_avail_output_samples = cd->chunk_frames * num_channels;
 
+	config = comp_get_data_blob(cd->tuning_handler, &config_size, NULL);
+
 	cd->state = GoogleCtcAudioProcessingCreateWithConfig(cd->chunk_frames,
 							     audio_stream_get_rate(&source->stream),
-							     /*config=*/NULL,
-							     /*config_size=*/0);
+							     config,
```
Comment (johnylin76) [inline src/audio/google/google_ctc_audio_processing.c:391]: Here should be better having the config size check (as you did in ctc_set_config()) before calling GoogleCtcAudioProcessingCreateWithConfig() in case a bad config blob is inhabiting. Consider to replace with {NULL, 0} for creating CTC under that circumstance.

Comment (eddy1021) [reply src/audio/google/google_ctc_audio_processing.c:396]: The config size is checked in `GoogleCtcAudioProcessingCreateWithConfig `. The config will be ignored if config size is not expected. I agree that checking again here would make it more clear.

**Review by lgirdwood — APPROVED** [review]
